# Car Pricing Prediction

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Project-description" data-toc-modified-id="Project-description-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Project description</a></span></li><li><span><a href="#Data-preperation" data-toc-modified-id="Data-preperation-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data preperation</a></span><ul class="toc-item"><li><span><a href="#Data-description" data-toc-modified-id="Data-description-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Data description</a></span></li><li><span><a href="#Removing-exces-data" data-toc-modified-id="Removing-exces-data-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Removing exces data</a></span></li><li><span><a href="#Data-validation" data-toc-modified-id="Data-validation-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Data validation</a></span><ul class="toc-item"><li><span><a href="#RegistrationYear" data-toc-modified-id="RegistrationYear-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>RegistrationYear</a></span></li><li><span><a href="#RegistrationMonth" data-toc-modified-id="RegistrationMonth-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>RegistrationMonth</a></span></li><li><span><a href="#Price" data-toc-modified-id="Price-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Price</a></span></li></ul></li><li><span><a href="#Missing-values-check." data-toc-modified-id="Missing-values-check.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Missing values check.</a></span><ul class="toc-item"><li><span><a href="#NotRepaired" data-toc-modified-id="NotRepaired-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>NotRepaired</a></span></li><li><span><a href="#VehicleType" data-toc-modified-id="VehicleType-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>VehicleType</a></span></li><li><span><a href="#FuelType" data-toc-modified-id="FuelType-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>FuelType</a></span></li><li><span><a href="#Gearbox" data-toc-modified-id="Gearbox-2.4.4"><span class="toc-item-num">2.4.4&nbsp;&nbsp;</span>Gearbox</a></span></li><li><span><a href="#Model" data-toc-modified-id="Model-2.4.5"><span class="toc-item-num">2.4.5&nbsp;&nbsp;</span>Model</a></span></li></ul></li><li><span><a href="#Encoding-categorical-data" data-toc-modified-id="Encoding-categorical-data-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Encoding categorical data</a></span><ul class="toc-item"><li><span><a href="#One-Hot-Encoding-(OHE)" data-toc-modified-id="One-Hot-Encoding-(OHE)-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>One-Hot Encoding (OHE)</a></span></li><li><span><a href="#Label-Encoding" data-toc-modified-id="Label-Encoding-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>Label Encoding</a></span></li></ul></li><li><span><a href="#Splitting-the-data" data-toc-modified-id="Splitting-the-data-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Splitting the data</a></span></li></ul></li><li><span><a href="#Model-training" data-toc-modified-id="Model-training-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Model training</a></span><ul class="toc-item"><li><span><a href="#Regular-model-training" data-toc-modified-id="Regular-model-training-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Regular model training</a></span><ul class="toc-item"><li><span><a href="#Linear-Regression" data-toc-modified-id="Linear-Regression-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Linear Regression</a></span></li><li><span><a href="#Decision-Tree" data-toc-modified-id="Decision-Tree-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Decision Tree</a></span></li><li><span><a href="#Random-Forest" data-toc-modified-id="Random-Forest-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Random Forest</a></span></li></ul></li><li><span><a href="#Gradient-Boosting" data-toc-modified-id="Gradient-Boosting-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Gradient Boosting</a></span><ul class="toc-item"><li><span><a href="#CatBoost" data-toc-modified-id="CatBoost-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>CatBoost</a></span></li><li><span><a href="#LightGBM" data-toc-modified-id="LightGBM-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>LightGBM</a></span></li><li><span><a href="#XGBoost" data-toc-modified-id="XGBoost-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>XGBoost</a></span></li><li><span><a href="#Testing-the-model" data-toc-modified-id="Testing-the-model-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Testing the model</a></span></li></ul></li></ul></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></div>

## Project description

Rusty Bargain used car sales service is developing an app to attract new customers. In that app, you can quickly find out the market value of your car. You have access to historical data: technical specifications, trim versions, and prices. You need to build the model to determine the value. 
Rusty Bargain is interested in:  

    •	the quality of the prediction  
    •	the speed of the prediction  
    •	the time required for training


In [1]:
# !pip install -U -q scikit-learn
# !pip install -U -q sidetable
# !pip install -U -q catboost
# !pip install -U -q lightgbm
# !pip install -U -q xgboost

In [2]:
import pandas as pd
import sidetable
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np
from scipy import stats as st
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error
from sklearn.metrics import make_scorer
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from catboost import CatBoostClassifier, Pool, cv
from catboost import CatBoostRegressor
import catboost as cb
import lightgbm as lgb
from xgboost.sklearn import XGBRegressor


from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## Data preparation

In [3]:
def load_data(file_name, parse_dates=None, dtype=None, sep=','):
    df=''
    
    try:
        df=pd.read_csv('datasets/{}'.format(file_name), parse_dates=parse_dates, dtype=dtype, sep=sep)
    except:
        df=pd.read_csv('/datasets/{}'.format(file_name), parse_dates=parse_dates, dtype=dtype, sep=sep)
        
    return df

Loading the data

In [4]:
df=load_data('car_data.csv')
df.head(2)

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power Model  \
0  24/03/2016 11:52    480         NaN              1993  manual      0  golf   
1  24/03/2016 10:58  18300       coupe              2011  manual    190   NaN   

   Mileage  RegistrationMonth  FuelType       Brand NotRepaired  \
0   150000                  0    petrol  volkswagen         NaN   
1   125000                  5  gasoline        audi         yes   

        DateCreated  NumberOfPictures  PostalCode          LastSeen  
0  24/03/2016 00:00                 0       70435  07/04/2016 03:16  
1  24/03/2016 00:00                 0       66954  07/04/2016 01:46

### Data description

__Features__

    DateCrawled — date profile was downloaded from the database
    VehicleType — vehicle body type
    RegistrationYear — vehicle registration year
    Gearbox — gearbox type
    Power — power (hp)
    Model — vehicle model
    Mileage — mileage (measured in km due to dataset's regional specifics)
    RegistrationMonth — vehicle registration month
    FuelType — fuel type
    Brand — vehicle brand
    NotRepaired — vehicle repaired or not
    DateCreated — date of profile creation
    NumberOfPictures — number of vehicle pictures
    PostalCode — postal code of profile owner (user)
    LastSeen — date of the last activity of the user

__Target__

    Price — price (Euro)

### Removing exces data

Some of the columns are not relevant to the model building process.  
All the data not related to cars will be removed.

In [5]:
df=df.drop(['DateCrawled', 'DateCreated', 'LastSeen', 'PostalCode', 'NumberOfPictures'], axis=1)

### Data validation

In [6]:
df.describe(include = 'all')

Price VehicleType  RegistrationYear Gearbox          Power  \
count   354369.000000      316879     354369.000000  334536  354369.000000   
unique            NaN           8               NaN       2            NaN   
top               NaN       sedan               NaN  manual            NaN   
freq              NaN       91457               NaN  268251            NaN   
mean      4416.656776         NaN       2004.234448     NaN     110.094337   
std       4514.158514         NaN         90.227958     NaN     189.850405   
min          0.000000         NaN       1000.000000     NaN       0.000000   
25%       1050.000000         NaN       1999.000000     NaN      69.000000   
50%       2700.000000         NaN       2003.000000     NaN     105.000000   
75%       6400.000000         NaN       2008.000000     NaN     143.000000   
max      20000.000000         NaN       9999.000000     NaN   20000.000000   

         Model        Mileage  RegistrationMonth FuelType       Brand  \
count   334664  354369.000000      354369.000000   321474      354369   
unique     250            NaN                NaN        7          40   
top       golf            NaN                NaN   petrol  volkswagen   
freq     29232            NaN                NaN   216352       77013   
mean       NaN  128211.172535           5.714645      NaN         NaN   
std        NaN   37905.341530           3.726421      NaN         NaN   
min        NaN    5000.000000           0.000000      NaN         NaN   
25%        NaN  125000.000000           3.000000      NaN         NaN   
50%        NaN  150000.000000           6.000000      NaN         NaN   
75%        NaN  150000.000000           9.000000      NaN         NaN   
max        NaN  150000.000000          12.000000      NaN         NaN   

       NotRepaired  
count       283215  
unique           2  
top             no  
freq        247161  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

#### RegistrationYear

In [7]:
df.sort_values('RegistrationYear').head(2)

Price VehicleType  RegistrationYear Gearbox  Power Model  Mileage  \
244092   2000         NaN              1000     NaN      0   NaN     5000   
91869     400         NaN              1000     NaN      0   NaN     5000   

        RegistrationMonth FuelType           Brand NotRepaired  
244092                  0      NaN  sonstige_autos         NaN  
91869                   0      NaN  sonstige_autos         NaN

The year 1000 is obviously a mistake.

In [8]:
cars_from_future=df.query('RegistrationYear>2016')
cars_from_future.head(2)

Price VehicleType  RegistrationYear Gearbox  Power   Model  Mileage  \
22   2900         NaN              2018  manual     90  meriva   150000   
26   5555         NaN              2017  manual    125      c4   125000   

    RegistrationMonth FuelType    Brand NotRepaired  
22                  5   petrol     opel          no  
26                  4      NaN  citroen          no

The data was created/collected in the year 2016.  
Cars newer than 2016 are from the future. It must be a mistake.  
So as cars older than 1900.

In [9]:
df.loc[df['RegistrationYear']>2016,'RegistrationYear']=np.NaN
df.loc[df['RegistrationYear']<1900,'RegistrationYear']=np.NaN

Those mistaken values will be filled in later while handling the rest of the missing values.

#### RegistrationMonth

In [10]:
df.RegistrationMonth.value_counts()

0     37352
3     34373
6     31508
4     29270
5     29153
7     27213
10    26099
12    24289
11    24186
9     23813
1     23219
8     22627
2     21267
Name: RegistrationMonth, dtype: int64

The dataframe has 13 months. Probably month 0 suppose to be january.

In [11]:
df.loc[df['RegistrationMonth']==0, 'RegistrationMonth']=1

#### Price

The price is the target. Rows that don't have a valid value are not relevant, not for trainning nor for testing.  
Rows with invalid price value will be removed.

In [12]:
pct_bad_price=len(df.query('Price==0'))/len(df)
print('The percent of invalild price tag is: {:.2%}'.format(pct_bad_price))

The percent of invalild price tag is: 3.04%


Deleting the rows with invalid price tag.

In [13]:
df.loc[df['Price']==0,'Price']=np.NaN
df=df.dropna(subset=['Price'])

### Missing values check.

In [14]:
df.stb.missing(style=True)

Since there are a lot of rows containing missing data, those rows can't be deleted.  
I'll handle each group of missing data according to its type.

#### NotRepaired

In [15]:
df.NotRepaired.value_counts()

no     243682
yes     34049
Name: NotRepaired, dtype: int64

In [16]:
pct_not_repaired=247161/(247161+36054)
print('Percent of not repaired: {:.2%}'.format(pct_not_repaired))

Percent of not repaired: 87.27%


Most of the observation's values of NotRepaired are no. I assume that most cars take it as obvious, and that's why this field is not filled up. Under this assumption, the missing values will be filled with no.

In [17]:
df.loc[df.NotRepaired.isna(), 'NotRepaired']='no'

#### VehicleType

In [18]:
df.VehicleType.value_counts()

sedan          89405
small          78018
wagon          63701
bus            28345
convertible    19865
coupe          15636
suv            11791
other           3086
Name: VehicleType, dtype: int64

There is no distinct value that can be used with high certainty as a filler to the missing values.

#### FuelType

In [19]:
df.FuelType.value_counts()

petrol      210765
gasoline     97168
lpg           5166
cng            554
hybrid         227
other          170
electric        88
Name: FuelType, dtype: int64

There is no consistency in name usage. I'll consolidate those groups.  
LPG and CNG are the same, the same goes with Petrol and Gasoline.  
Fuel type other will be redefined as petrol like most of the cars.

In [20]:
df.loc[df['FuelType']=='gasoline','FuelType']='petrol'
df.loc[df['FuelType']=='cng','FuelType']='lpg'
df.loc[df['FuelType']=='other','FuelType']='petrol'

df.FuelType.value_counts()

petrol      308103
lpg           5720
hybrid         227
electric        88
Name: FuelType, dtype: int64

In [21]:
print('Petrol cars are {:.2%} of all observations'.format(df.FuelType.value_counts()['petrol']/len(df)))

Petrol cars are 89.67% of all observations


Since more than 89% of all observations are petrol cars, I'll label the fuel type missing values as petrol.

In [22]:
df.loc[df['FuelType'].isna(),'FuelType']='petrol'

#### Gearbox

In [23]:
df.Gearbox.value_counts()

manual    261373
auto       64869
Name: Gearbox, dtype: int64

In [24]:
print('Petrol cars are {:.2%} of all observations'.format(df.Gearbox.value_counts()['manual']/len(df)))

Petrol cars are 76.07% of all observations


Since more than 76% of all observation's gearbox type is manual, I'll label the gearbox type missing values as manual.

In [25]:
df.loc[df['Gearbox'].isna(), 'Gearbox']='manual'

#### Model

In [26]:
df.Model.value_counts()

golf                  28246
other                 23707
3er                   19179
polo                  12628
corsa                 12197
                      ...  
serie_2                   8
rangerover                4
serie_3                   3
range_rover_evoque        2
serie_1                   1
Name: Model, Length: 250, dtype: int64

There is no distinct value that can be used with high certainty as a filler to the missing values.

After treating all kinds of data issues, and some of the missing values, those are the currently missing values:

In [27]:
df.stb.missing(style=True)

In [28]:
df_no_na=df.dropna()
msg='There are {} observations without any missing values, which are {:.2%} from the original dataset.'
print(msg.format(len(df_no_na), len(df_no_na)/ len(df)))

There are 297892 observations without any missing values, which are 86.70% from the original dataset.


The missing data that can't be filled with high certainty, will be deleted.

In [29]:
df=df_no_na

The price feature is the target for our models.

In [30]:
X_df=df.drop('Price', axis=1)
y_df=df.Price

### Encoding categorical data

Categorical data needs to be converted to numerical values. The requirement is to build models using different algorithms (random forest, decision tree, and linear regression). The linear regression works better with OHE, while the decision tree works better with label encoding.  
The 2 different methods will be applied and used according to the algorithm's needs.

#### One-Hot Encoding (OHE) 

Encoding using OHE.

In [31]:
X_df_ohe=pd.get_dummies(X_df, drop_first=True)
X_df_ohe.head()

RegistrationYear  Power  Mileage  RegistrationMonth  \
2            2004.0    163   125000                  8   
3            2001.0     75   150000                  6   
4            2008.0     69    90000                  7   
5            1995.0    102   150000                 10   
6            2004.0    109   150000                  8   

   VehicleType_convertible  VehicleType_coupe  VehicleType_other  \
2                        0                  0                  0   
3                        0                  0                  0   
4                        0                  0                  0   
5                        0                  0                  0   
6                        1                  0                  0   

   VehicleType_sedan  VehicleType_small  VehicleType_suv  ...  Brand_seat  \
2                  0                  0                1  ...           0   
3                  0                  1                0  ...           0   
4                  0                  1                0  ...           0   
5                  1                  0                0  ...           0   
6                  0                  0                0  ...           0   

   Brand_skoda  Brand_smart  Brand_subaru  Brand_suzuki  Brand_toyota  \
2            0            0             0             0             0   
3            0            0             0             0             0   
4            1            0             0             0             0   
5            0            0             0             0             0   
6            0            0             0             0             0   

   Brand_trabant  Brand_volkswagen  Brand_volvo  NotRepaired_yes  
2              0                 0            0                0  
3              0                 1            0                0  
4              0                 0            0                0  
5              0                 0            0                1  
6              0                 0            0                0  

[5 rows x 303 columns]

#### Label Encoding

Encoding using label encoding technique.  
Finding the categorical columns.

In [58]:
numerical_features=[]
categorical_features=[]
for col in list(X_df.columns):
    if X_df[col].dtypes == object:
        categorical_features.append(col)
    else:
        numerical_features.append(col)
        
X_df[categorical_features].head(2)
X_df[numerical_features].head(2)

VehicleType Gearbox  Model FuelType       Brand NotRepaired
2         suv    auto  grand   petrol        jeep          no
3       small  manual   golf   petrol  volkswagen          no

RegistrationYear  Power  Mileage  RegistrationMonth
2            2004.0    163   125000                  8
3            2001.0     75   150000                  6

Label encoding is applied only to the categorical columns.

In [60]:
# column_transformer needs to have the columns that needs to be transformed first,
# and only then the columns that won't be transformed - numerical features
features_reordered=categorical_features+numerical_features
X_df=X_df[features_reordered]
        
X_df_labeled=X_df.copy()
column_transformer=make_column_transformer(
    (OrdinalEncoder(), categorical_features),
    remainder='passthrough')
X_df_labeled[:]=column_transformer.fit_transform(X_df)
X_df_labeled.head()

VehicleType  Gearbox  Model  FuelType  Brand  NotRepaired  \
2          6.0      0.0  117.0       3.0   14.0          0.0   
3          5.0      1.0  116.0       3.0   37.0          0.0   
4          5.0      1.0  101.0       3.0   31.0          0.0   
5          4.0      1.0   11.0       3.0    2.0          1.0   
6          1.0      1.0    8.0       3.0   25.0          0.0   

   RegistrationYear  Power   Mileage  RegistrationMonth  
2            2004.0  163.0  125000.0                8.0  
3            2001.0   75.0  150000.0                6.0  
4            2008.0   69.0   90000.0                7.0  
5            1995.0  102.0  150000.0               10.0  
6            2004.0  109.0  150000.0                8.0

### Splitting the data

Splitting the data for training and testing datasets.

In [42]:
X_train_lbl, X_test_lbl, y_train_lbl, y_test_lbl = train_test_split(
    X_df_labeled, y_df, test_size=0.3, random_state=12345
)

In [43]:
X_train_ohe, X_test_ohe, y_train_ohe, y_test_ohe = train_test_split(
    X_df_ohe, y_df, test_size=0.3, random_state=12345
)

## Model training

The requested metric is __RMSE__. A function is defined for that, and a scorer to use this self-made metric with the `GridSearchCV()`.  
The model training compares the result of standard models:  

    Linear Regression, Decision Tree, Random Forest
    
to Gradient Boosting algorithms:  

    CatBoost, LightGBM, XBoost

__Feature scaling__ - Making sure that all features should be considered equally important before the algorithm's execution. In order to avoid data leakage, pipeline will be used.

In [47]:
def RMSE(y_true, y_pred):
    return mean_squared_error(y_true, y_pred, squared=False)

In [92]:
RMSE_scorer=make_scorer(RMSE, greater_is_better=False)
std_scaler = StandardScaler()

def search_best_hyper_params(model, params, X, y):
    
    pipe=make_pipeline(std_scaler, model)
    
    clf=GridSearchCV(pipe, param_grid=params, scoring=RMSE_scorer, cv=10)
    clf.fit(X, y)
    return clf.best_score_*(-1), clf.best_params_

### Regular model training

#### Linear Regression

In [95]:
%%time
lr_model=LinearRegression()
search_best_hyper_params(lr_model, {}, X_train_ohe, y_train_ohe)

Wall time: 58.3 s


(2728.632616355445, {})

Training time for 200K observations:

In [98]:
%%time
lr_pipe=make_pipeline(std_scaler, lr_model)
lr_pipe.fit(X_train_ohe, y_train_ohe)

Wall time: 5.41 s


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('linearregression', LinearRegression())])

Prediction time for 200K observations:

In [99]:
%%time
lr_pipe.predict(X_train_ohe)

Wall time: 571 ms


array([15538.78994385, -2798.51905026,  1169.74717977, ...,
        2979.34785228,  1336.88696769,  9610.57039248])

#### Decision Tree

In [53]:
%%time
search_best_hyper_params(
    DecisionTreeRegressor(random_state=12345), 
    {
        'decisiontreeregressor__max_features':['auto','log2','sqrt',None],
        'decisiontreeregressor__max_leaf_nodes':[None,10,90],
        'decisiontreeregressor__max_depth':[2, 3, 4]
    },
    X_train_lbl, y_train_lbl
)

Wall time: 52.5 s


(2514.4083364190005,
 {'decisiontreeregressor__max_depth': 4,
  'decisiontreeregressor__max_features': 'auto',
  'decisiontreeregressor__max_leaf_nodes': None})

Training time using selected hyperparameters for 200K observations:

In [100]:
%%time
dt_model=DecisionTreeRegressor(random_state=12345, max_features='auto', max_leaf_nodes=None, max_depth=4)
dt_pipe=make_pipeline(std_scaler, dt_model)
dt_pipe.fit(X_train_lbl, y_train_lbl)

Wall time: 262 ms


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('decisiontreeregressor',
                 DecisionTreeRegressor(max_depth=4, max_features='auto',
                                       random_state=12345))])

Prediction time for 200K observations

In [101]:
%%time
dt_pipe.predict(X_train_lbl)

Wall time: 37.9 ms


array([14705.93219214,  6138.58252835,  1596.73819281, ...,
        4113.5408805 ,  1596.73819281, 10669.74552286])

#### Random Forest

In [54]:
%%time
search_best_hyper_params(
    RandomForestRegressor(random_state=12345), 
    {'randomforestregressor__n_estimators':[20, 80], 
     'randomforestregressor__max_depth':[3, 6] },
    X_train_lbl, y_train_lbl
)

Wall time: 5min 13s


(2206.4292558000357,
 {'randomforestregressor__max_depth': 6,
  'randomforestregressor__n_estimators': 80})

Training time using selected hyperparameters for 200K observations:

In [102]:
%%time
rf_model=RandomForestRegressor(random_state=12345, n_estimators=80, max_depth=6)
rf_pipe=make_pipeline(std_scaler, rf_model)
rf_pipe.fit(X_train_lbl, y_train_lbl)

Wall time: 19 s


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('randomforestregressor',
                 RandomForestRegressor(max_depth=6, n_estimators=80,
                                       random_state=12345))])

Prediction time for 200K observations

In [103]:
%%time
rf_pipe.predict(X_train_lbl)

Wall time: 965 ms


array([13451.90470796,  4278.09327521,  1775.82467675, ...,
        4233.05884516,  1183.11186333,  9257.51232789])

Tried 3 regular algorithms with some parameters:
    
    Linear Regression
    Decision Tree
    Random Forest

The best result was received from the Random Forest with RMSE score of 2206.43.  
Training time using selected hyperparameters for 200K observations took 19 seconds, and the prediction time was 965 milliseconds

### Gradient Boosting

Running the same data using Gradient Boosting algorithms.

#### CatBoost

In [69]:
%%time
search_best_hyper_params(
    cb.CatBoostRegressor(
        loss_function='RMSE', random_seed = 12345, verbose=False
    ), 
    {'catboostregressor__iterations': [100, 200, 500],
     'catboostregressor__learning_rate': [0.03, 0.1],
     'catboostregressor__depth': [2, 5, 8]
    },
    X_train_lbl, y_train_lbl
)

Custom logger is already specified. Specify more than one logger at same time is not thread safe.

Wall time: 18min 1s


(1596.0519301517338,
 {'catboostregressor__depth': 8,
  'catboostregressor__iterations': 500,
  'catboostregressor__learning_rate': 0.1})

Training time using selected hyperparameters for 200K observations:

In [104]:
%%time
cb_model=cb.CatBoostRegressor(
        loss_function='RMSE', random_seed = 12345, verbose=False, 
        iterations=500, depth=8, learning_rate=0.1
    )
cb_pipe=make_pipeline(std_scaler, cb_model)
cb_pipe.fit(X_train_lbl, y_train_lbl)

Wall time: 14.9 s


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('catboostregressor',
                 <catboost.core.CatBoostRegressor object at 0x000001AB5A3A0D60>)])

Prediction time for 200K observations

In [105]:
%%time
cb_pipe.predict(X_train_lbl)

Wall time: 111 ms


array([  920.09214186,  1444.95991244,  1037.80737925, ...,
        3272.17909191,  1023.01175918, 11440.19279029])

The CatBoost algorithm managed to outperform the best score a regular model managed to get. The Random Forest got an RMSE score of 2206.43, while the CatBoost got an impressive score of 1596.05.  
Training time using selected hyperparameters for 200K observations took 14.9 seconds, and the prediction time was 111 milliseconds

#### LightGBM

In [55]:
%%time
search_best_hyper_params(
    lgb.LGBMRegressor(random_state=12345), 
    {
        'lgbmregressor__learning_rate': [0.03, 0.1],
        'lgbmregressor__max_depth': [2, 5, 8],
        'lgbmregressor__n_estimators': [100, 200, 500]
    },
    X_train_lbl, y_train_lbl
)

Wall time: 6min 8s


(1580.2131350122704,
 {'lgbmregressor__learning_rate': 0.1,
  'lgbmregressor__max_depth': 8,
  'lgbmregressor__n_estimators': 500})

Training time using selected hyperparameters for 200K observations:

In [106]:
%%time
lgbm_model=lgb.LGBMRegressor(random_state=12345, learning_rate=0.1, max_depth=8, n_estimators=500)
lgbm_pipe=make_pipeline(std_scaler, lgbm_model)
lgbm_pipe.fit(X_train_lbl, y_train_lbl)

Wall time: 2.83 s


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('lgbmregressor',
                 LGBMRegressor(max_depth=8, n_estimators=500,
                               random_state=12345))])

Prediction time for 200K observations

In [107]:
%%time
lgbm_pipe.predict(X_train_lbl)

Wall time: 2.34 s


array([ 1404.44719291,  2313.29111321,   832.3513787 , ...,
        3211.61884185,  1053.76487258, 10907.03991158])

The LightGBM algorithm managed to get a slightly better score than the CatBoost algorithm. The LightGBM scored 1580.21, while CatBoost scored 1596.05. Using similar parameters, the training time was slightly faster. The training time of LightGBM was 2.83 seconds, while the CatBoost did it in 14.9 seconds.

#### XGBoost

In [70]:
%%time

search_best_hyper_params(
    XGBRegressor(random_state=12345), 
    {
      'xgbregressor__nthread':[4],
      'xgbregressor__learning_rate': [0.03, 0.1],
      'xgbregressor__max_depth': [2, 8],
      'xgbregressor__n_estimators': [100, 500]
    },
    X_train_lbl, y_train_lbl
)

Wall time: 31min 41s


(1528.416488117165,
 {'xgbregressor__learning_rate': 0.1,
  'xgbregressor__max_depth': 8,
  'xgbregressor__n_estimators': 500,
  'xgbregressor__nthread': 4})

Training time using selected hyperparameters for 200K observations:

In [108]:
%%time
xgb_model=XGBRegressor(random_state=12345, nthread=4, learning_rate=0.1, max_depth=8, n_estimators=500)
xgb_pipe=make_pipeline(std_scaler, xgb_model)
xgb_pipe.fit(X_train_lbl, y_train_lbl)

Wall time: 1min 7s


Pipeline(steps=[('standardscaler', StandardScaler()),
                ('xgbregressor',
                 XGBRegressor(base_score=0.5, booster='gbtree',
                              colsample_bylevel=1, colsample_bynode=1,
                              colsample_bytree=1, enable_categorical=False,
                              gamma=0, gpu_id=-1, importance_type=None,
                              interaction_constraints='', learning_rate=0.1,
                              max_delta_step=0, max_depth=8, min_child_weight=1,
                              missing=nan, monotone_constraints='()',
                              n_estimators=500, n_jobs=4, nthread=4,
                              num_parallel_tree=1, predictor='auto',
                              random_state=12345, reg_alpha=0, reg_lambda=1,
                              scale_pos_weight=1, subsample=1,
                              tree_method='exact', validate_parameters=1,
                              verbosity=None))])

Prediction time for 200K observations

In [109]:
%%time
xgb_pipe.predict(X_train_lbl)

Wall time: 1.5 s


array([ 1293.0513,  1319.6216,   918.1772, ...,  3505.5098,  1014.9354,
       11145.523 ], dtype=float32)

The XGBoost did manage to perform slightly better than the rest of the gradient boosting algorithms using the same parameters, it got a score of 1528.41 compared to 1580.21 and 1596.05, but it consumed much more time (1 minute while the other managed to do it in 2.83 seconds for the LightGBM, and in 14.9 seconds in the CatBoost).

#### Testing the model

After checking all those algorithms, I'll choose the LightGBM as it gave very good results, and in a very good time compared to the other gradient boosting algorithms.

In [110]:
lgbm_model=lgb.LGBMRegressor(random_state=12345, learning_rate=0.1, max_depth=8, n_estimators=500)
lgbm_pipe=make_pipeline(std_scaler, lgbm_model)
lgbm_pipe.fit(X_train_lbl, y_train_lbl)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('lgbmregressor',
                 LGBMRegressor(max_depth=8, n_estimators=500,
                               random_state=12345))])

In [111]:
lgbm_pred = lgbm_pipe.predict(X_test_lbl)
rmse_score = (RMSE(y_test_lbl, lgbm_pred))

print('LightGBM RMSE: {:.2f}'.format(rmse_score))

LightGBM RMSE: 1564.68


The score of the test dataset for the LightGBM model was slightly better than the score in the training phase (1564.68 compared to 1580.21), which means that the training was good, and it was not overfitting.

## Conclusion

Algorithms using gradient boosting techniques are much more accurate than traditional algorithms. Using the same parameters the CatBoost, LightGBM, and XGBoost got similar RMSE scores, while the XGBoost was slower than the other 2. The CatBoost and LightGBM had similar results and training time, but the LightGBM was slightly better, and it is the winner from all of the algorithms tested here.